data/transactions/logic/eval_test.go

```diff
@@ -5117,3 +5117,93 @@ func TestOpBase64Decode(t *testing.T) {
 	args := b64TestDecodeAssembleWithArgs(t)
 	b64TestDecodeEval(t, args)
 }
+
+func TestOpJSONRef(t *testing.T) {
+	partitiontest.PartitionTest(t)
+	t.Parallel()
+	proto := defaultEvalProtoWithVersion(LogicVersion)
+	txn := transactions.SignedTxn{
+		Txn: transactions.Transaction{
+			Type: protocol.ApplicationCallTx,
+		},
+	}
+	ledger := logictest.MakeLedger(nil)
+	ledger.NewApp(txn.Txn.Receiver, 0, basics.AppParams{})
+	sb := strings.Builder{}
+	ep := defaultEvalParams(&sb, &txn)
+	ep.Proto = &proto
+	ep.Ledger = ledger
+	testCases := []struct {
+		source string
+	}{
+		{
+			source: `byte  "{\"key0\": 1,\"key1\": \"algo\",\"key2\":{\"key3\": \"teal\", \"key3\": [1,2,3]} }"; byte "key0"; json_ref JSONInt; int 1; ==`,
+		},
+		{
+			source: `byte  "{\"key0\": 1,\"key1\": \"algo\",\"key2\":{\"key3\": \"teal\", \"key3\": [1,2,3]} }"; byte "key1"; json_ref JSONString; byte "algo"; ==`,
+		},
+		{
+			source: `byte  "{\"key0\": 1,\"key1\": \"algo\",\"key2\":{\"key3\": \"teal\", \"key4\": [1,2,3]} }"; byte "key2.key3"; json_ref JSONString; byte "teal"; ==`,
+		},
+		{
+			source: `byte  "{\"rawId\": \"responseId\",\"id\": \"0\",\"response\": {\"attestationObject\": \"based64url_encoded_buffer\",\"clientDataJSON\":  \"based64url_encoded_client_data\"},\"getClientExtensionResults\": {},\"type\": \"public-key\"}"; byte "response.clientDataJSON"; json_ref JSONString; byte "based64url_encoded_client_data"; ==`,
+		},
```

Comment:
We need to test negative numbers, and floats, which should both fail. (If you try to read a value and it's -5, or 3.2, it should not silently coerce it to some integer). We also need to test very larger integers, close to 2^64.  In particular, we need to have tests that ensure an integer that can not be represented well by a float gets treated properly without conversion into and out of floating point.

I imagine we have to ask Go to parse using big.Int to represent the outcome.

Reply:
Can be addressed by this comment: https://github.com/algorand/go-algorand/pull/3349/files#r776080968